lib/data_stores/mysql_advanced_data_store.py

```diff
@@ -0,0 +1,636 @@
+      self._ExecuteQuery(query, args)
+
+  def DeleteAttributesRegex(self, subject, regexes, token=None):
+    self.security_manager.CheckDataStoreAccess(token, [subject], "w")
+
+    for regex in regexes:
+      query, args = self._BuildQuery("DELETE", subject, regex, is_regex=True)
+      self._ExecuteQuery(query, args)
+
+  def DeleteSubject(self, subject, token=None, sync=False):
+    _ = sync
+    self.security_manager.CheckDataStoreAccess(token, [subject], "w")
+
+    query, args = self._BuildQuery("DELETE", subject)
+    self._ExecuteQuery(query, args)
+
+  def ResolveMulti(self, subject, predicates, token=None, timestamp=None):
+    """Resolves multiple predicates at once for one subject."""
+    self.security_manager.CheckDataStoreAccess(
+        token, [subject], self.GetRequiredResolveAccess(predicates))
+
+    for predicate in predicates:
+      query, args = self._BuildQuery("SELECT", subject, predicate, timestamp)
+      result = self._ExecuteQuery(query, args)
+
+      for row in result:
+        value = self._DecodeValue(predicate, row["value"])
+
+        yield predicate, value, rdfvalue.RDFDatetime(row["timestamp"])
+
+  def MultiResolveRegex(self, subjects, predicate_regex, token=None, timestamp=None, limit=None):
+    """Result multiple subjects using one or more predicate regexps."""
+    self.security_manager.CheckDataStoreAccess(
+        token, subjects, self.GetRequiredResolveAccess(predicate_regex))
+
+    result = {}
+
+    for subject in subjects:
+      values = self._ResolveRegex(subject, predicate_regex, token=token,
+                                  timestamp=timestamp, limit=limit)
+
+      if values:
+        result[subject] = values
+        if limit:
+          limit -= len(values)
+
+      if limit and limit <= 0:
+        break
+
+    return result.iteritems()
+
+  def MultiSet(self, subject, values, timestamp=None, token=None, replace=True,
+               sync=True, to_delete=None):
+    """Set multiple predicates' values for this subject in one operation."""
+    self.security_manager.CheckDataStoreAccess(token, [subject], "w")
+    to_delete = set(to_delete or [])
+    if timestamp is None:
+      timestamp = time.time() * 1e6
+
+    # Prepare a bulk insert operation.
+    subject = utils.SmartUnicode(subject)
+    to_insert = []
+    to_replace = []
+
+    # Build a document for each unique timestamp.
+    for attribute, sequence in values.items():
+      for value in sequence:
+        entry_timestamp = None
+
+        if isinstance(value, tuple):
+          value, entry_timestamp = value
+
+        if entry_timestamp is None:
+          entry_timestamp = timestamp
+
+        predicate = utils.SmartUnicode(attribute)
+        data = self._EncodeValue(value)
+
+        # Replacing means to delete all versions of the attribute first.
+        if replace or attribute in to_delete:
+          duplicates = self._CountDuplicateAttributes(subject, predicate)
```

Comment:
The more I think about this the more I suspect I might actually be able to change this.  Since there is no ORDER BY involved in this so MySQL should be able to use the indexes to get the results back quickly.  I will see what I can do on this.